src/components/Rating/Rating.tsx

```diff
@@ -0,0 +1,201 @@
+
+  // managing classnames
+  let styleClasses = classNames('rating_icon-group', {
+    disabled: !!disabled,
+    readonly: !!readonly,
+  });
+  if (className) {
+    styleClasses += ' ' + className;
+  }
+  // rating array as the default to stars state
+  const starArr = new Array(ratingCount).fill(0).map((item, index) => {
+    return {
+      value: 0,
+      index: index + 1,
+    };
+  });
+
+  // states to manage save current rating value,currentHover position, and stars value
+  const [currentTotalRating, setCurrentTotalRating] = useState(defaultRating);
+  const [currentHovering, setCurrentHovering] = useState<number>(-1);
+  const [stars, setStars] = useState<IStars>(starArr);
+
+  // custom hook to track the current rating label
+  const { labelName } = useCurrentRatingLabel(
+    currentTotalRating as number,
+    ratingCount as number
+  );
+
+  const handleOnClick = () => {
+    if (disabled || readonly) {
+      return;
+    }
+    // get current total star rating
+    const newStarRating = getCurrentStarRating(stars);
+    setCurrentTotalRating(newStarRating);
+    // onChange callback
+    if (onChange) {
+      onChange(currentTotalRating);
+    }
+  };
+
+  // reset animation effect
+  const handleOnMouseLeave = () => {
+    if (disabled || readonly) {
+      return;
+    }
+    setCurrentHovering(-1);
+  };
+
+  // hovering effect with half and full stars preview
+  const handleOnMouseMove = (e: any, index: number) => {
+    if (disabled || readonly) {
+      return;
+    }
+    // set current hovering position
+    setCurrentHovering(index);
+    // get current star ratings
+    const newStarRating = getCurrentStarRating(stars);
+    setCurrentTotalRating(newStarRating);
+    // half stars logic
+    if (half) {
+      const { offsetX } = e.nativeEvent;
+      const currentSize = convertSizeNameToSizeNumber(size as IconSize);
+      const newStars = getCurrentHalfStars(offsetX, stars, index, currentSize);
+      setStars(newStars);
+    } else {
+      const newStars = getFullStars(stars, index);
+      setStars(newStars);
+    }
+  };
+
+  //check defaultRating value and half prop to update stars
+  useEffect(() => {
+    if (defaultRating && !Number.isInteger(defaultRating) && half) {
+      const newStars = getDefaultHalfStars(stars, defaultRating);
+      setStars(newStars);
+    } else {
+      const newStars = getFullStars(stars, defaultRating as number);
+      setStars(newStars);
+    }
+  }, [defaultRating, half]);
+
+  //Rating component to be rendered
+  const renderedRating = stars.map((star, index) => {
+    const ratingValue = index + 1;
+    const currentSize = getSizeName(
+      size as IconSize,
+      currentHovering === ratingValue
+    );
+    return (
+      <div
+        key={star.index + 'star'}
+        className={styleClasses}
+        onMouseLeave={handleOnMouseLeave}
+        onMouseMove={(e) => handleOnMouseMove(e, ratingValue)}
+        onClick={handleOnClick}
+        role="button"
```

Comment:
It is because I am using role to select the element in testing. `getByRole` requires an attribute called `role = 'aria role'` in order to select the element for testing.